clippy_lints/src/map_contains_then_remove.rs

```diff
@@ -0,0 +1,96 @@
+use rustc::hir::*;
+use rustc::hir::def::Def;
+use rustc::hir::intravisit::{Visitor, NestedVisitorMap, walk_expr};
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass, LintContext};
+use syntax::ast::NodeId;
+use utils::span_help_and_lint;
+
+/// **What it does:** TODO
+///
+/// **Why is this bad?** TODO
+///
+/// **Known problems:** TODO
+///
+/// **Example:** TODO
+
+declare_lint!{
+    pub MAP_CONTAINS_THEN_REMOVE,
+    Warn,
+    "TODO:description"
+}
+
+struct MapRemoveVisitor<'a, 'tcx: 'a> {
+    cx: &'a LateContext<'a, 'tcx>,
+    map_id: NodeId,
+    key_id: NodeId,
+    removed: bool,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for MapRemoveVisitor<'a, 'tcx> {
+    fn visit_expr(&mut self, expr: &'tcx Expr) {
+        if let ExprMethodCall(ref method, _, ref args) = expr.node {
+            if method.name == "remove" && args.len() == 2 {
+                self.cx.sess().span_note_without_error(expr.span, "found a candidate, now checking if map/key IDs match");
+                if same_var(self.cx, &args[0], self.map_id) && same_var(self.cx, &args[1], self.key_id) {
```

Comment:
So, for the map this makes sense, since you want to check if the same map is modified. For the argument this doesn't quite work in your test cases. `same_var` checks whether it's the same variable. `&1` is neither a variable, nor has `&1` the same address as another `&1`. What you actually want is to compare the expressions and check if they are "essentially" the same. Search for `eq_expr` in the clippy repository to see some examples of how to do that.

Reply:
Thanks! In fact the check didn't work for map either (but did work once revised to check local_id against local_id, rather than local_id against node_id), but I've replaced it in my local branch with a check using eq_expr and now it works for both args. I'll clean up the lint and update the PR soon.